API.md

```diff
@@ -0,0 +1,337 @@
+# MemoryStreamBufferType Enum
+
+Namespace: Microsoft.IO
+
+
+## Values
+
+| Name | Summary |
+|---|---|
+| **Small** |  |
+| **Large** |  |
+# MemoryStreamDiscardReason Enum
+
+Namespace: Microsoft.IO
+
+
+## Values
+
+| Name | Summary |
+|---|---|
+| **TooLarge** |  |
+| **EnoughFree** |  |
+# RecyclableMemoryStream Class
+
+Namespace: Microsoft.IO
+
+Base class: MemoryStream
+
+MemoryStream implementation that deals with pooling and managing memory streams which use potentially large
+buffers.
+
+## Remarks
+
+This class works in tandem with the RecyclableMemoryStreamManager to supply MemoryStream
+objects to callers, while avoiding these specific problems:
+1. LOH allocations - since all large buffers are pooled, they will never incur a Gen2 GC
+2. Memory waste - A standard memory stream doubles its size when it runs out of room. This
+leads to continual memory growth as each stream approaches the maximum allowed size.
+3. Memory copying - Each time a MemoryStream grows, all the bytes are copied into new buffers.
+This implementation only copies the bytes when GetBuffer is called.
+4. Memory fragmentation - By using homogeneous buffer sizes, it ensures that blocks of memory
+can be easily reused.
+
+The stream is implemented on top of a series of uniformly-sized blocks. As the stream's length grows,
+additional blocks are retrieved from the memory manager. It is these blocks that are pooled, not the stream
+object itself.
+
+The biggest wrinkle in this implementation is when GetBuffer() is called. This requires a single 
+contiguous buffer. If only a single block is in use, then that block is returned. If multiple blocks 
+are in use, we retrieve a larger buffer from the memory manager. These large buffers are also pooled, 
+split by size--they are multiples/exponentials of a chunk size (1 MB by default).
+
+Once a large buffer is assigned to the stream the blocks are NEVER again used for this stream. All operations take place on the 
+large buffer. The large buffer can be replaced by a larger buffer from the pool as needed. All blocks and large buffers 
+are maintained in the stream until the stream is disposed (unless AggressiveBufferReturn is enabled in the stream manager).
+
+## Properties
+
+| Name | Type | Summary |
+|---|---|---|
+| **Id** | Guid | Unique identifier for this stream across it's entire lifetime |
+| **Tag** | string | A temporary identifier for the current usage of this stream. |
+| **MemoryManager** | [RecyclableMemoryStreamManager](#recyclablememorystreammanager-class) | Gets the memory manager being used by this stream. |
+| **AllocationStack** | string | Callstack of the constructor. It is only set if MemoryManager.GenerateCallStacks is true,<br>which should only be in debugging situations. |
+| **DisposeStack** | string | Callstack of the Dispose call. It is only set if MemoryManager.GenerateCallStacks is true,<br>which should only be in debugging situations. |
+| **Capacity** | int | Gets or sets the capacity |
```

Comment:
how can capacity be an int and length a long?  Also why would I ever set this?  what happens if I write past the capacity?

Reply:
Unfortunately, the types are determined by the base class, MemoryStream, and I'm adhering to those. I think it makes no sense, but it is what it is.

I'll add some documentation about exceeding the capacity somewhere.